error "context provider not set" in log

## Bug Report

Please answer these questions before submitting your issue. Thanks!

### 1. Minimal reproduce step (Required)

<!-- a step by step guide for reproducing the bug. -->
found in an internal test log

### 2. What did you expect to see? (Required)

### 3. What did you see instead (Required)

```
[2025/01/08 21:45:33.514 +08:00] [INFO] [job_worker.go:387] ["finish DDL job"] [category=ddl] [jobID=215] [conn=3407880844] [job="ID:215, Type:drop table, State:synced, SchemaState:none, SchemaID:134, TableID:212, RowCount:0, ArgLen:1, start time: 2025-01-08 21:45:33.381 +0800 CST, Err:<nil>, ErrCount:0, SnapshotVersion:0, Version: v2"]
[2025/01/08 21:45:33.567 +08:00] [INFO] [txn.go:55] ["Try to create a new txn inside a transaction auto commit"] [category=ddl-notifier] [schemaVersion=178] [txnStartTS=455172144111550531] [txnScope=global]
[2025/01/08 21:45:33.592 +08:00] [ERROR] [txn.go:721] ["txnManager failed to handle OnStmtRollback"] [error="context provider not set"]
```

### 4. What is your TiDB version? (Required)

<!-- Paste the output of SELECT tidb_version() -->

6c9ee36e8e7c86abbda374d841b019979ff3aaf4

--- a/pkg/ddl/notifier/BUILD.bazel
+++ b/pkg/ddl/notifier/BUILD.bazel
@@ -36,7 +36,7 @@ go_test(
     ],
     embed = [":notifier"],
     flaky = True,
-    shard_count = 9,
+    shard_count = 10,
     deps = [
         "//pkg/ddl",
         "//pkg/ddl/session",

--- a/pkg/ddl/notifier/subscribe.go
+++ b/pkg/ddl/notifier/subscribe.go
@@ -235,14 +235,15 @@ func (n *DDLNotifier) processEvents(ctx context.Context) error {
 				if err3 != nil {
 					return errors.Trace(err3)
 				}
-				defer n.sysSessionPool.Put(s3)
 				sess4Del := sess.NewSession(s3.(sessionctx.Context))
-				if err3 = n.store.DeleteAndCommit(
+				err3 = n.store.DeleteAndCommit(
 					ctx,
 					sess4Del,
 					change.ddlJobID,
 					int(change.subJobID),
-				); err3 != nil {
+				)
+				n.sysSessionPool.Put(s3)
+				if err3 != nil {
 					logutil.Logger(ctx).Error("Error deleting change",
 						zap.Int64("ddlJobID", change.ddlJobID),
 						zap.Int64("subJobID", change.subJobID),

--- a/pkg/ddl/notifier/subscribe.go
+++ b/pkg/ddl/notifier/subscribe.go
@@ -231,9 +231,15 @@ func (n *DDLNotifier) processEvents(ctx context.Context) error {
 			}
 
 			if change.processedByFlag == n.handlersBitMap {
-				if err3 := n.store.DeleteAndCommit(
+				s3, err3 := n.sysSessionPool.Get()
+				if err3 != nil {
+					return errors.Trace(err3)
+				}
+				defer n.sysSessionPool.Put(s3)
+				sess4Del := sess.NewSession(s3.(sessionctx.Context))
+				if err3 = n.store.DeleteAndCommit(
 					ctx,
-					sess4List,
+					sess4Del,
 					change.ddlJobID,
 					int(change.subJobID),
 				); err3 != nil {

--- a/pkg/ddl/notifier/testkit_test.go
+++ b/pkg/ddl/notifier/testkit_test.go
@@ -108,12 +108,7 @@ func TestBasicPubSub(t *testing.T) {
 		return nil
 	}
 	n.RegisterHandler(notifier.TestHandlerID, testHandler)
-
-	done := make(chan struct{})
-	go func() {
-		n.OnBecomeOwner()
-		close(done)
-	}()
+	n.OnBecomeOwner()
 
 	tk2 := testkit.NewTestKit(t, store)
 	se := sess.NewSession(tk2.Session())
@@ -138,7 +133,6 @@ func TestBasicPubSub(t *testing.T) {
 	require.Equal(t, event2, seenChanges[1])
 	require.Equal(t, event3, seenChanges[2])
 	n.OnRetireOwner()
-	<-done
 }
 
 func TestDeliverOrderAndCleanup(t *testing.T) {
@@ -186,12 +180,7 @@ func TestDeliverOrderAndCleanup(t *testing.T) {
 	n.RegisterHandler(3, h1)
 	n.RegisterHandler(4, h2)
 	n.RegisterHandler(9, h3)
-
-	done := make(chan struct{})
-	go func() {
-		n.OnBecomeOwner()
-		close(done)
-	}()
+	n.OnBecomeOwner()
 
 	tk2 := testkit.NewTestKit(t, store)
 	se := sess.NewSession(tk2.Session())
@@ -220,7 +209,6 @@ func TestDeliverOrderAndCleanup(t *testing.T) {
 	require.Equal(t, []int64{1000, 1001, 1002}, *id3)
 
 	n.OnRetireOwner()
-	<-done
 }
 
 func TestPubSub(t *testing.T) {
@@ -352,12 +340,7 @@ func Test2OwnerForAShortTime(t *testing.T) {
 		return nil
 	}
 	n.RegisterHandler(notifier.TestHandlerID, testHandler)
-
-	done := make(chan struct{})
-	go func() {
-		n.OnBecomeOwner()
-		close(done)
-	}()
+	n.OnBecomeOwner()
 
 	tk2 := testkit.NewTestKit(t, store)
 	se := sess.NewSession(tk2.Session())
@@ -383,7 +366,6 @@ func Test2OwnerForAShortTime(t *testing.T) {
 	tk2.MustQuery("SELECT * FROM test.result").Check(testkit.Rows())
 
 	n.OnRetireOwner()
-	<-done
 }
 
 func TestPaginatedList(t *testing.T) {
@@ -451,4 +433,59 @@ func TestPaginatedList(t *testing.T) {
 	require.Eventually(t, func() bool {
 		return count.Load() == 8
 	}, 5*time.Second, 500*time.Millisecond)
+}
+
+func TestBeginTwice(t *testing.T) {
+	conf := new(log.Config)
+	logFilename := path.Join(t.TempDir(), "/testBeginTwice.log")
+	conf.File.Filename = logFilename
+	lg, p, e := log.InitLogger(conf)
+	require.NoError(t, e)
+	rs := log.ReplaceGlobals(lg, p)
+	defer rs()
+
+	store := testkit.CreateMockStore(t)
+	tk := testkit.NewTestKit(t, store)
+	tk.MustExec("USE test")
+	tk.MustExec("DROP TABLE IF EXISTS " + ddl.NotifierTableName)
+	tk.MustExec(ddl.NotifierTableSQL)
+
+	s := notifier.OpenTableStore("test", ddl.NotifierTableName)
+	sessionPool := util.NewSessionPool(
+		5,
+		func() (pools.Resource, error) {
+			return testkit.NewTestKit(t, store).Session(), nil
+		},
+		nil,
+		nil,
+	)
+
+	n := notifier.NewDDLNotifier(sessionPool, s, 50*time.Millisecond)
+
+	testHandler := func(context.Context, sessionctx.Context, *notifier.SchemaChangeEvent) error {
+		return nil
+	}
+	n.RegisterHandler(notifier.TestHandlerID, testHandler)
+	n.OnBecomeOwner()
+
+	tk2 := testkit.NewTestKit(t, store)
+	se := sess.NewSession(tk2.Session())
+	ctx := context.Background()
+	event1 := notifier.NewCreateTableEvent(&model.TableInfo{ID: 1000, Name: ast.NewCIStr("t1")})
+	err := notifier.PubSchemeChangeToStore(ctx, se, 1, -1, event1, s)
+	require.NoError(t, err)
+
+	// after handler processed the event, wait to ensure the record is deleted by DDL notifier
+	require.Eventually(t, func() bool {
+		changes := make([]*notifier.SchemaChange, 8)
+		result, closeFn := s.List(ctx, se)
+		count, err2 := result.Read(changes)
+		require.NoError(t, err2)
+		closeFn()
+		return count == 0
+	}, time.Second, 50*time.Millisecond)
+
+	content, err := os.ReadFile(logFilename)
+	require.NoError(t, err)
+	require.NotContains(t, string(content), "context provider not set")
 }

